### PR TITLE
(PUP-7579) Allow tags to contain Unicode words

### PR DESCRIPTION
**Pull request by Iristyle**
- The regex for tags was originally /^\w[-\w:]*$/ in a Luke commit
    from 10 years ago f98be4a7198326b26f1072c401b3e337f340db40

  - Over the years it has received updates to modernize it and has
    shuffled around.

  - 0badfce72bafc767bf3099718c63324f51108996 changed it from

    /^\w[-\w:.]*$/

    to

    /^[0-9A-Za-z_][0-9A-Za-z_:.-]*$/

    for performance reasons.

  - The concat module depends on tags, and as such requires that tags
    can contain other valid content.  The regex here is too
    restrictive for that purpose and should be returned to its
    original intent - word characters.

    However, it now needs to be updated to accept word characters in
    any language.

  - This now allows for the follow manifest to not cause an error:

  puppet apply -e 'notify { "hi": tag => "_メインディレクトリ_醸造所"  }'

### CODE REVIEW

**Comment by Iristyle** [comment]
Would be nice to have a few tests that ensured no regressions.

File: lib/puppet/util/tagging.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 require 'puppet/util/tag_set'
 
 module Puppet::Util::Tagging
-  ValidTagRegex = /\A[0-9A-Za-z_][0-9A-Za-z_:.-]*\Z/
+  ValidTagRegex = /\A[[:word:]][[[:word:]]:.-]*\Z/
```
Comment (thallgren) [inline lib/puppet/util/tagging.rb]: I'm not overly fond of [:word:] since it's a non-standard construct. Documentation states that it includes everything in the general Unicode category 'Mark' and that doesn't seems right. Also, I think there are one pair of brackets too many in the expression. How about this?
```
/\A[[:alnum:]_][[:alnum:]_:.-]*\z/
```

Comment (thallgren) [inline lib/puppet/util/tagging.rb]: Please add tests.

Comment (Iristyle) [inline lib/puppet/util/tagging.rb]: The `[[` / `]]` brackets look weird, but are correct - see https://ruby-doc.org/core-2.1.1/Regexp.html#class-Regexp-label-Character+Classes

I can give `:alnum:` a shot with some test data to see what happens.  We're mostly concerned with making tags compatible with the concat module for https://tickets.puppetlabs.com/browse/MODULES-4864.

Comment (thallgren) [inline lib/puppet/util/tagging.rb]: The extra brackets are redundant. The example you point at is already a character group. You don't need an extra set of brackets if you're already in one when the objective is to add more characters to the same character group.
The appointed ticket seems somewhat unrelated.

Comment (Iristyle) [inline lib/puppet/util/tagging.rb]: Ah ah, you were referring to the `[[[:word:]]:.-]` being rewritable as `[[:word:]:.-]` - I see. 
 Yes, that's fine.  

I thought you were referring to the first `[[:word:]]` which cannot have the double brackets removed: - i.e. 

```
[1] pry(main)> "abc" =~ /[:word:]/
=> nil
[2] pry(main)> "abc" =~ /[[:word:]]/
=> 0
```

As for the modules ticket, I've been told that concat makes extensive use of tags, and without tag names fully supporting UTF-8, concat explodes - so this is the root issue (at least with how concat is currently implemented).  I haven't looked at the concat code nor am I aware if it can be reasonably rewritten without tags.

Comment (thallgren) [inline lib/puppet/util/tagging.rb]: OK, but that's probably [PUP-4684](https://tickets.puppetlabs.com/browse/MODULES-4684), not 4864 ;-)

Comment (Iristyle) [inline lib/puppet/util/tagging.rb]: Yes!  Thanks

https://tickets.puppetlabs.com/browse/MODULES-4684

Comment (joshcooper) [inline lib/puppet/util/tagging.rb]: We'll need to update docs: https://docs.puppet.com/puppet/4.10/lang_reserved.html#tags

**Comment by MosesMendoza** [comment]
PuppetDB also appears to validate tags against a similar schema: https://github.com/puppetlabs/puppetdb/blob/master/src/puppetlabs/puppetdb/catalogs.clj#L304-L320

File: lib/puppet/util/tagging.rb (hunk 1 above)
Comment (Iristyle) [inline lib/puppet/util/tagging.rb]: With @MosesMendoza confirming that PuppetDB code uses RegEx for tag validation as well, I agree that we should not use the Ruby `[[:word:]]` @thallgren and should instead rely on the POSIX compatible `[[:alnum:]]`, which has a Java equivalent in https://docs.oracle.com/javase/6/docs/api/java/util/regex/Pattern.html as `\p{Alnum}`.

/cc @mullr @rbrw

**Comment by Iristyle** [comment]
PR has been updated and should be ready for merge.  @puppetlabs/puppetdb should make corresponding changes to coincide with these.

**Comment by Iristyle** [comment]
@MFredette if you could point me to the docs I can put up a pull request there as well.  Thanks!

**Comment by joshcooper** [comment]
Is there a PDB ticket?

**Comment by Iristyle** [comment]
Looks like the documentation updates should land at https://github.com/puppetlabs/puppet-docs/blob/master/source/puppet/5.0/lang_reserved.markdown

**Comment by Iristyle** [comment]
Update doc PR in #6016
